src/lib.rs

```diff
@@ -8,34 +8,93 @@ use core::cmp::Ordering;
 
 use proc_macro2::{Ident, TokenStream};
 use quote::{format_ident, quote, ToTokens};
-use syn::parse::{Parse, ParseStream};
-use syn::punctuated::Punctuated;
-use syn::spanned::Spanned;
 use syn::{
-    Data, DeriveInput, Error, Fields, FieldsNamed, FieldsUnnamed, Result, Token, TraitBound, Type,
+    parse::{discouraged::Speculative, Parse, ParseStream},
+    punctuated::Punctuated,
+    spanned::Spanned,
+    Data, DeriveInput, Error, Fields, FieldsNamed, FieldsUnnamed, PredicateType, Result, Token,
+    Type, WherePredicate,
 };
 
-/// Holds parsed [bounds](TraitBound) and [traits](Traits).
-struct DeriveWhere {
-    /// Parsed [bounds](TraitBound).
-    bounds: Vec<TraitBound>,
-    /// Parsed [traits](Traits).
-    traits: Vec<Traits>,
+/// Holds a single generic [type](Type) or [type with bound](PredicateType)
+enum Generic {
+    /// Generic type with custom [specified bounds](PredicateType)
+    CoustomBound(PredicateType),
+    /// Generic [type](Type) which will be bound by the implemented trait
+    NoBound(Type),
 }
 
-impl Parse for DeriveWhere {
+impl Parse for Generic {
     fn parse(input: ParseStream) -> Result<Self> {
-        let bounds = Punctuated::<TraitBound, Token![,]>::parse_separated_nonempty(input)?
-            .into_iter()
-            .collect();
+        let fork = input.fork();
+        // Try to parse input as a WherePredicate. The problem is, both expresions
+        // start with a Type, so this is the easiest way of differenciating them.
+        match WherePredicate::parse(&fork) {
+            Ok(where_predicate) => {
+                // Advance input as if `WherePredicate` was parsed on it.
+                input.advance_to(&fork);
+
+                match where_predicate {
+                    WherePredicate::Type(type_) => Ok(Self::CoustomBound(type_)),
+                    WherePredicate::Lifetime(_) => Err(Error::new(
+                        where_predicate.span(),
+                        "Bounds on lifetimes are not supported",
+                    )),
+                    WherePredicate::Eq(_) => Err(Error::new(
+                        where_predicate.span(),
+                        "Equality predicates are not supported",
+                    )),
+                }
+            }
+            Err(_) => Ok(Self::NoBound(input.parse()?)),
+        }
+    }
+}
 
-        <Token![;]>::parse(input)?;
+/// Holds parsed [generics](Generic) and [traits](Traits).
+enum DeriveWhere {
+    /// Generic type parameters were defined.
+    WithGenerics(Vec<Generic>, Vec<Traits>),
+    /// Only traits were set.
+    OnlyTraits(Vec<Traits>),
+}
```

Comment:
```suggestion
struct DeriveWhere {
  traits: Vec<Traits>,
  generics: Option<VecGeneric>>,
```
This seems more rusty to me, what do you think?

Reply:
Don't really have a preference